languageserver/src/main/scala/langserver/core/TextDocument.scala

```diff
@@ -0,0 +1,88 @@
+package langserver.core
+
+import langserver.types.TextDocumentIdentifier
+import langserver.types.TextDocumentContentChangeEvent
+import langserver.types.Position
+import java.io.File
+import java.net.URI
+
+case class TextDocument(uri: String, contents: Array[Char]) {
+  def applyChanges(changes: Seq[TextDocumentContentChangeEvent]): TextDocument = {
+    // we assume full text sync
+    assert(changes.size == 1)
+    val change = changes.head
+    assert(change.range.isEmpty)
+    assert(change.rangeLength.isEmpty)
+
+    copy(contents = change.text.toArray)
+  }
+
+  private def peek(idx: Int) =
+    if (idx < contents.size) contents(idx) else -1
+
+  def toFile: File =
+    new File(URI.create(uri))
+
+  /**
+   * Return the corresponding position in this text document as 0-based line and column.
+   */
+  def offsetToPosition(offset: Int): Position = {
+    if (offset >= contents.size)
+      throw new IndexOutOfBoundsException(s"$uri: asked position at offset $offset, but contents is only ${contents.size} characters long.")
+
+    var i, line, col = 0
+
+    while (i < offset) {
+      contents(i) match {
+        case '\r' =>
+          line += 1
+          col = 0
+          if (peek(i + 1) == '\n') i += 1
+
+        case '\n' =>
+          line += 1
+          col = 0
+
+        case _ =>
+          col += 1
+      }
+      i += 1
+    }
+
+    Position(line, col)
+  }
+
+  /**
+   * Return the offset in the current document, for a given 0-based line/col position.
+   */
+  def positionToOffset(pos: Position): Int = {
```

Comment:
I hadn't noticed this was implemented, we should use it instead of the buggy implementation I added on monday.

Reply:
Ah nice, it's the first time I notice it too!